index.js

```diff
@@ -11,420 +10,432 @@ const omit = require('lodash/omit')
-  const TableName = req.params.TableName
-  describeTable({TableName}).then((description) => {
-    const params = {
-      TableName,
-      Item: req.body
-    }
+    }).catch(next)
+  })
 
-    return putItem(params).then((response) => {
-      const Key = extractKey(req.body, description.Table.KeySchema)
+  app.get('/tables/:TableName/items/:key', (req, res, next) => {
+    const TableName = req.params.TableName
+    describeTable({TableName}).then((result) => {
       const params = {
         TableName,
-        Key
+        Key: parseKey(req.params.key, result.Table)
       }
+
       return getItem(params).then((response) => {
         if (!response.Item) {
           return res.status(404).send('Not found')
         }
-        return res.json(Key)
+        res.render('item', {
+          TableName: req.params.TableName,
+          Item: response.Item,
+          isNew: false
+        })
       })
-    })
-  }).catch(next)
-})
-
-app.put('/tables/:TableName/items/:key', bodyParser.json(), (req, res, next) => {
-  const TableName = req.params.TableName
-  describeTable({TableName}).then((result) => {
-    const params = {
-      TableName,
-      Item: req.body
-    }
+    }).catch(next)
+  })
 
-    return putItem(params).then(() => {
+  app.put('/tables/:TableName/add-item', bodyParser.json(), (req, res, next) => {
+    const TableName = req.params.TableName
+    describeTable({TableName}).then((description) => {
       const params = {
         TableName,
-        Key: parseKey(req.params.key, result.Table)
+        Item: req.body
       }
-      return getItem(params).then((response) => {
-        return res.json(response.Item)
+
+      return putItem(params).then((response) => {
+        const Key = extractKey(req.body, description.Table.KeySchema)
+        const params = {
+          TableName,
+          Key
+        }
+        return getItem(params).then((response) => {
+          if (!response.Item) {
+            return res.status(404).send('Not found')
+          }
+          return res.json(Key)
+        })
       })
-    })
-  }).catch(next)
-})
+    }).catch(next)
+  })
 
-const port = process.env.PORT || 8001
-app.listen(port, () => {
-  console.log(`dynamodb-admin listening on port ${port}`)
-})
+  app.put('/tables/:TableName/items/:key', bodyParser.json(), (req, res, next) => {
+    const TableName = req.params.TableName
+    describeTable({TableName}).then((result) => {
+      const params = {
+        TableName,
+        Item: req.body
+      }
+
+      return putItem(params).then(() => {
+        const params = {
+          TableName,
+          Key: parseKey(req.params.key, result.Table)
+        }
+        return getItem(params).then((response) => {
+          return res.json(response.Item)
+        })
+      })
+    }).catch(next)
+  })
+
+  return app
+}
+
+module.exports = buildServer
+
+if (runningFromCli) {
+  const env = process.env;
```

Comment:
This `environment` configuration should be executed when not running from cli too, otherwise we won't be able to run local dynamodb-admin from `module`

Reply:
Perhaps we can export a function to auto-configure in this way. I am very hesitant to execute this code when this module is required by another, because that other module might have already configured the AWS object (which is global!) and clobber its state. I'm very much against `require()`ing modules causing global state to change.

Reply:
Perhaps I can put it another way.  If you are using this module in another module, you are probably using AWS services in that other module/application already.  I would expect that application to have already configured the global AWS object sufficiently for its purposes, and any kind of auto-config this module might do (1) shouldn't be necessary in the first place, and (2) could very easily step on the configuration that the including module already did.

Reply:
Got your thought and I totally agree with you.
As both [DynamoDb](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/DynamoDB.html#constructor-property) and [DynamoDb.DocumentClient](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/DynamoDB/DocumentClient.html#constructor-property) constructors accepts configuration parameters we can set these configurations only for the local instances used by `dynamodb-admin` instead of putting it into the global `AWS` instance.
This way we don't risk overwriting our global AWS configuration that may be set from another module

Reply:
Yup, that was my thought when I added those arguments.  We construct our DynamoDB and DocumentClients with options, so it seemed to make sense to allow the caller to provide them.

Incidentally, this is why I substituted `AWS` with `require('aws-sdk')` wherever it is used.  My thought was that if you provide a mock object for DynamoDB and DocumentClient, there's no reason to require `aws-sdk` at all, so this module now only requires it when it's clear that it's necessary.

Reply:
`so it seemed to make sense to allow the caller to provide them` absolutely, the caller might already have both instances, and passing them as parameters would save some bytes in memory, and also have better performance, since it doesn't need to create a new instance every time.

`there's no reason to require aws-sdk at all, so this module now only requires it when it's clear that it's necessary.` that's ok.

But in this case, the caller MUST have these instances pointing to a local dynamo and pass them as arguments to the function to connect to a local dynamo, setting a environment variable looks a lot easier. And as this project works only by just setting an env var, i think we should support this functionality when `require`ing it too.

So if no dynamo/docClient instances are provided, imo, they should be created with "default" parameters read from environment.

Reply:
> And as this project works only by just setting an env var, i think we should support this functionality when `require`ing it too.

As long as that's not modifying global state, I would be okay with that -- reading settings from the environment variable and then creating the DynamoDB and document client with those settings.

But I'm not okay with the global `AWS` object being reconfigured in that case.

Reply:
We're on the same page then!
Polluting `AWS` global object must be avoided.

Can you update tour PR to create dynamo/docClient to create these objects with parameters from environment if the objects are not provided ? Then we're ready to merge.

Reply:
Sorry for the delay, things got really busy for me.

I've fixed this and updated the PR.